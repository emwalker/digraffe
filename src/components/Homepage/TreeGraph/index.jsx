// @flow
import React, { Component } from 'react'
import { ForceGraph3D } from 'react-force-graph'

const nodeLabel = d => `<div style="text-align: center; padding: 8px; color: black; border-radius:
    5px; background: rgba(239, 243, 255, 0.80); font-size: 28">${d.name}</div>`

const onNodeClick = (node) => {
  const url = `/wiki/topics/${node.id}`
  try {
    window.open(url)
  } catch (e) {
    window.location.href = url
  }
}

type Props = {
  height: number,
  topicGraph: string,
  width: number,
}

class TreeGraph extends Component<Props> {
  cachedGraphData: ?string = null

  graphRef: ?any

  componentDidMount = () => {
    if (this.graphRef) this.graphRef.d3Force('charge').strength(-100)
  }

  get graphData(): ?string {
    if (!this.props.topicGraph) return null
    if (!this.cachedGraphData) this.cachedGraphData = JSON.parse(this.props.topicGraph)

    return this.cachedGraphData
  }

  render = () => (
    <ForceGraph3D
      backgroundColor="white"
      dagLevelDistance={90}
      dagMode="td"
      graphData={this.graphData}
      height={this.props.height}
      linkColor={() => 'rgba(49, 83, 160, 0.3)'}
      linkWidth={3}
      nodeColor={() => 'rgb(82, 97, 140)'}
      nodeLabel={nodeLabel}
      nodeOpacity={1}
      nodeResolution={15}
      nodeVal={d => d.topicCount}
      onNodeClick={onNodeClick}
      ref={(ref) => { this.graphRef = ref }}
      showNavInfo={false}
      width={this.props.width}
    />
  )
}

export default TreeGraph
