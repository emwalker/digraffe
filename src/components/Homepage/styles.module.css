.container {
  max-width: 720px;
  margin: auto;
}

.container > p {
  margin-bottom: 3em;
}

.list {
  margin-bottom: 3em;
}

.search {
  margin-top: 5em !important;
}

.hero {
  width: 51%;
  text-align: center;
  margin: auto auto 4em auto;
}

@media (max-width: 1012px) {
  .hero {
    width: 95%;
  }
}

.logo, .logo:hover {
  text-decoration: none;
  color: inherit;
}
