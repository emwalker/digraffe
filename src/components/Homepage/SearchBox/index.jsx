// @flow
import React, { Component } from 'react'
import classNames from 'classnames'

import { everythingTopicPath } from 'components/constants'
import styles from './styles.module.css'

type Props = {
  className?: ?string,
  router: {
    push: Function,
  },
}

class SearchBox extends Component<Props> {
  static defaultProps = {
    className: '',
  }

  onKeyPress = (event: SyntheticKeyboardEvent<HTMLButtonElement>) => {
    if (event.key === 'Enter') {
      const { value } = (event.target: window.HTMLInputElement)
      this.onSearch(value)
    }
  }

  onSearch = (query: string) => {
    if (query === '') {
      this.props.router.push({ pathname: everythingTopicPath })
      return
    }

    this.props.router.push({ pathname: everythingTopicPath, query: { q: query } })
  }

  render = () => (
    <p className={classNames(this.props.className, 'form-group text-center')}>
      <input
        aria-label="Search"
        className={classNames(styles.input, 'form-control p-3')}
        onKeyPress={this.onKeyPress}
        placeholder="Start with a search"
        type="search"
      />
    </p>
  )
}

export default SearchBox
