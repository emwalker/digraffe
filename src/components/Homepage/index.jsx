import React from 'react'
import { graphql } from 'react-relay'
import classNames from 'classnames'

import { Homepage_homepage_QueryResponse as Response } from './__generated__/Homepage_homepage_Query.graphql'
import LineItem from './LineItem'
import SearchBox from './SearchBox'
import styles from './styles.module.css'

type ViewType = $PropertyType<Response, 'view'>

type Props = {
  router: Object,
  view: ViewType,
}

const Homepage = ({ view, router }: Props) => (
  <div
    className={classNames(styles.container, 'px-3 px-md-6 px-lg-0')}
  >
    <h2 className="mb-2">
      Digraph
    </h2>

    <ul className={classNames(styles.list, 'ml-4 f4')}>
      <li>Save links in a mind mapâ€“like network of topics.</li>
      <li>Keep track of everything you&apos;ve read or might want to read in the future.</li>
      <li>
        Gain control over your reading and turn the flood of information into knowledge.
      </li>
    </ul>

    <h4>Recent updates</h4>
    <div className="f4">
      {view.activity.edges.map(({ node }) => <LineItem key={node.description} item={node} />)}

      <div>
        There are currently
        {` ${view.linkCount.toLocaleString()} `}
        links and
        {` ${view.topicCount.toLocaleString()} `}
        topics.
      </div>
    </div>

    <SearchBox className={styles.search} router={router} />
  </div>
)

export const query = graphql`
query Homepage_homepage_Query(
  $viewerId: ID!,
  $orgLogin: String!,
  $repoName: String,
  $repoIds: [ID!],
) {
  view(
    viewerId: $viewerId,
    currentOrganizationLogin: $orgLogin,
    currentRepositoryName: $repoName,
    repositoryIds: $repoIds,
  ) {
    linkCount
    topicCount

    activity(first: 3) {
      edges {
        node {
          description
          ...LineItem_item
        }
      }
    }
  }
}`

export default Homepage
