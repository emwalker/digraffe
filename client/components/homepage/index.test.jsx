import React from 'react'
import { shallow } from 'enzyme'
import Homepage from './index'

describe('<Homepage />', () => {
  const wrapper = shallow(<Homepage />)

  it('renders', () => {
    expect(wrapper).toMatchSnapshot()
  })
})
