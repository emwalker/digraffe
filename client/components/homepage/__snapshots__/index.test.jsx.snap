// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Homepage /> renders 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Unknown />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <HelmetWrapper
          defer={true}
          encodeSpecialCharacters={true}
          meta={
            Array [
              Object {
                "content": "Golang Isomorphic React/Hot Reloadable/Redux/Css-Modules Starter Kit",
                "property": "og:title",
              },
            ]
          }
          title="Home page"
        />,
        <h1
          className="example"
        >
          Hot Reloadable 
          <br />
          Golang + React + Redux + Css-Modules
          <br />
          Isomorphic Starter Kit
        </h1>,
        <br />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "defer": true,
          "encodeSpecialCharacters": true,
          "meta": Array [
            Object {
              "content": "Golang Isomorphic React/Hot Reloadable/Redux/Css-Modules Starter Kit",
              "property": "og:title",
            },
          ],
          "title": "Home page",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Hot Reloadable ",
            <br />,
            "Golang + React + Redux + Css-Modules",
            <br />,
            "Isomorphic Starter Kit",
          ],
          "className": "example",
        },
        "ref": null,
        "rendered": Array [
          "Hot Reloadable ",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": "br",
          },
          "Golang + React + Redux + Css-Modules",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": "br",
          },
          "Isomorphic Starter Kit",
        ],
        "type": "h1",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": "br",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <HelmetWrapper
            defer={true}
            encodeSpecialCharacters={true}
            meta={
              Array [
                Object {
                  "content": "Golang Isomorphic React/Hot Reloadable/Redux/Css-Modules Starter Kit",
                  "property": "og:title",
                },
              ]
            }
            title="Home page"
          />,
          <h1
            className="example"
          >
            Hot Reloadable 
            <br />
            Golang + React + Redux + Css-Modules
            <br />
            Isomorphic Starter Kit
          </h1>,
          <br />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "defer": true,
            "encodeSpecialCharacters": true,
            "meta": Array [
              Object {
                "content": "Golang Isomorphic React/Hot Reloadable/Redux/Css-Modules Starter Kit",
                "property": "og:title",
              },
            ],
            "title": "Home page",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Hot Reloadable ",
              <br />,
              "Golang + React + Redux + Css-Modules",
              <br />,
              "Isomorphic Starter Kit",
            ],
            "className": "example",
          },
          "ref": null,
          "rendered": Array [
            "Hot Reloadable ",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": "br",
            },
            "Golang + React + Redux + Css-Modules",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": "br",
            },
            "Isomorphic Starter Kit",
          ],
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "br",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenAdapter {
      "options": Object {
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
  },
}
`;
